File: lib/src/features/celebrations/celebrations.dart
import 'package:flutter_riverpod/flutter_riverpod.dart';

import '../members/application/member_providers.dart';
import '../members/domain/member.dart';
import '../members/domain/relationship.dart';

enum CelebrationKind { birthday, anniversary }

/// An upcoming birthday or wedding anniversary.
class Celebration {
  const Celebration({
    required this.kind,
    required this.title,
    required this.date,
    required this.daysUntil,
    required this.years,
    this.memberId,
    this.photoUrl,
  });

  final CelebrationKind kind;
  final String title;
  final DateTime date; // next occurrence
  final int daysUntil;
  final int years; // age turning, or years married (0 if unknown)
  final String? memberId; // person to greet (birthdays)
  final String? photoUrl;

  bool get isToday => daysUntil == 0;
}

/// Next occurrence of [date]'s month/day on or after [from] (date-only).
DateTime _nextOccurrence(DateTime date, DateTime from) {
  final today = DateTime(from.year, from.month, from.day);
  var next = DateTime(today.year, date.month, date.day);
  if (next.isBefore(today)) {
    next = DateTime(today.year + 1, date.month, date.day);
  }
  return next;
}

/// Upcoming celebrations within the next [withinDays] days, soonest first.
final upcomingCelebrationsProvider =
    FutureProvider.family<List<Celebration>, String>((ref, familyId) async {
  final members = await ref.watch(membersProvider(familyId).future);
  final relationships =
      await ref.watch(relationshipsProvider(familyId).future);
  return computeCelebrations(members, relationships);
});

/// Pure function (testable) that derives celebrations from members + edges.
List<Celebration> computeCelebrations(
  List<Member> members,
  List<Relationship> relationships, {
  int withinDays = 60,
  DateTime? now,
}) {
  final today = now ?? DateTime.now();
  final byId = {for (final m in members) m.id: m};
  final result = <Celebration>[];

  // Birthdays — living members with a known birth date.
  for (final m in members) {
    if (!m.isLiving || m.birthDate == null) continue;
    final next = _nextOccurrence(m.birthDate!, today);
    final days = next.difference(DateTime(today.year, today.month, today.day)).inDays;
    if (days <= withinDays) {
      result.add(Celebration(
        kind: CelebrationKind.birthday,
        title: m.fullName,
        date: next,
        daysUntil: days,
        years: next.year - m.birthDate!.year,
        memberId: m.id,
        photoUrl: m.photoUrl,
      ));
    }
  }

  // Anniversaries — spouse/partner unions with a known start date.
  final seen = <String>{};
  for (final r in relationships) {
    if (!r.isUnion || r.startDate == null) continue;
    final a = byId[r.fromMember];
    final b = byId[r.toMember];
    if (a == null || b == null) continue;
    final key = a.id.compareTo(b.id) < 0 ? '${a.id}|${b.id}' : '${b.id}|${a.id}';
    if (!seen.add(key)) continue;
    final next = _nextOccurrence(r.startDate!, today);
    final days = next.difference(DateTime(today.year, today.month, today.day)).inDays;
    if (days <= withinDays) {
      result.add(Celebration(
        kind: CelebrationKind.anniversary,
        title: '${a.firstName} & ${b.firstName}',
        date: next,
        daysUntil: days,
        years: next.year - r.startDate!.year,
      ));
    }
  }

  result.sort((x, y) => x.daysUntil.compareTo(y.daysUntil));
  return result;
}

File: webapp/src/lib/celebrations.ts
import type { Member, Relationship } from './types'
import { fullName } from './member-utils'

export interface Celebration {
  kind: 'birthday' | 'anniversary'
  date: Date // next occurrence
  daysUntil: number
  title: string
  subtitle: string
  turning?: number // age / years married
  memberIds: string[]
}

function nextOccurrence(month: number, day: number, from: Date): { date: Date; days: number } {
  const year = from.getFullYear()
  let d = new Date(year, month, day)
  const today = new Date(from.getFullYear(), from.getMonth(), from.getDate())
  if (d < today) d = new Date(year + 1, month, day)
  const days = Math.round((d.getTime() - today.getTime()) / 86_400_000)
  return { date: d, days }
}

/** Upcoming birthdays (living members) + wedding anniversaries (active unions). */
export function upcomingCelebrations(
  members: Member[],
  rels: Relationship[],
  withinDays = 366,
): Celebration[] {
  const now = new Date()
  const byId = new Map(members.map((m) => [m.id, m]))
  const out: Celebration[] = []

  for (const m of members) {
    if (!m.is_living || !m.birth_date) continue
    const b = new Date(m.birth_date)
    if (isNaN(b.getTime())) continue
    const { date, days } = nextOccurrence(b.getMonth(), b.getDate(), now)
    if (days > withinDays) continue
    out.push({
      kind: 'birthday',
      date,
      daysUntil: days,
      title: `${fullName(m)}’s birthday`,
      subtitle: date.toLocaleDateString(undefined, { month: 'long', day: 'numeric' }),
      turning: date.getFullYear() - b.getFullYear(),
      memberIds: [m.id],
    })
  }

  for (const r of rels) {
    if (r.type !== 'spouse' && r.type !== 'partner') continue
    if (!r.start_date || r.end_date) continue
    const s = new Date(r.start_date)
    if (isNaN(s.getTime())) continue
    const a = byId.get(r.from_member)
    const b = byId.get(r.to_member)
    if (!a || !b) continue
    const { date, days } = nextOccurrence(s.getMonth(), s.getDate(), now)
    if (days > withinDays) continue
    out.push({
      kind: 'anniversary',
      date,
      daysUntil: days,
      title: `${a.first_name} & ${b.first_name}’s anniversary`,
      subtitle: date.toLocaleDateString(undefined, { month: 'long', day: 'numeric' }),
      turning: date.getFullYear() - s.getFullYear(),
      memberIds: [a.id, b.id],
    })
  }

  return out.sort((x, y) => x.daysUntil - y.daysUntil)
}
